modules/dns_common/dns_common.rb

```diff
@@ -20,6 +20,77 @@ def resolver
       Resolv::DNS.new(:nameserver => @server)
     end
 
+    def create_a_record(fqdn, ip)
+      case a_record_conflicts(fqdn, ip) #returns -1, 0, 1
+        when 1 then
+          raise(Proxy::Dns::Collision, "'#{fqdn} 'is already in use")
+        when 0 then
+          return nil
+        else
+          do_create(fqdn, ip, "A")
+      end
+    end
+
+    def create_aaaa_record(fqdn, ip)
+      case aaaa_record_conflicts(fqdn, ip) #returns -1, 0, 1
+        when 1 then
+          raise(Proxy::Dns::Collision, "'#{fqdn} 'is already in use")
+        when 0 then
+          return nil
+        else
+          do_create(fqdn, ip, "AAAA")
+      end
+    end
+
+    def create_cname_record(fqdn, target)
+      case cname_record_conflicts(fqdn, target) #returns -1, 0, 1
+        when 1 then
+          raise(Proxy::Dns::Collision, "'#{fqdn} 'is already in use")
+        when 0 then
+          return nil
+        else
+          do_create(fqdn, target, "CNAME")
+      end
+    end
+
+    def create_ptr_record(fqdn, ptr)
+      case ptr_record_conflicts(fqdn, ptr) #returns -1, 0, 1
+        when 1 then
+          raise(Proxy::Dns::Collision, "'#{ptr}' is already in use")
+        when 0
+          return nil
+        else
+          do_create(ptr, fqdn, "PTR")
+      end
+    end
+
+    def do_create(name, value, type)
+      raise(Proxy::Dns::Error, "Creation of #{type} not implemented")
```

Comment:
I know this is what we have for do_remove, but is it actually that useful? By default ruby will throw "undefined method" exception, while this will result in a somewhat misleading error message (if you are getting this message do_create isn't implemented at all, not just for a particular record type).

I think having a test that validates that provider implements a required interface might be more helpful (for an example pls. see https://github.com/theforeman/smart-proxy/blob/develop/test/dhcp_isc/dhcp_isc_provider_interface_test.rb)?